### PR TITLE
Create initial reference implementation

### PR DESCRIPTION
**Pull request by benlesh**
Just a spike in TypeScript. Will add tests in a followup, add compilation to JavaScript, etc.

### CODE REVIEW

**Comment by benlesh** [comment]
Could even just convert it to JavaScript. I'm relatively indifferent.

File: impl/observable.ts [hunk 1]
```diff
@@ -0,0 +1,496 @@
+/**
+MIT License
```
Comment (domfarolino) [inline impl/observable.ts:2]: Hmm, should we perhaps make this a separate LICENSE file in the top-level of the repo?

Comment (benlesh) [reply impl/observable.ts:2]: I think that's fine. Mostly I like licensing things because of the liability.

**Comment by benlesh** [comment]
@domfarolino should we merge this? is it useful?

File: impl/observable.js [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+				});
+			},
+			[Symbol.asyncIterator]() {
+				return this;
+			},
+		};
+	}
+}
+class Subscriber {
+	#active = true;
```
Comment (gitawego) [inline impl/observable.js:553]: just a question, do we really need to use private properties ? as it's not widely available in browsers.

Comment (bakkot) [reply impl/observable.js:553]: Private fields have been available in evergreen browsers for more than two years. https://caniuse.com/mdn-javascript_classes_private_class_fields

Comment (ljharb) [reply impl/observable.js:553]: Also, this isn't a polyfill, it's a reference implementation. Polyfills should generally strive for maximal backwards compatibility, but reference implementations should strive for the most modern standard syntax and APIs available.

File: LICENSE.txt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2023 Ben Lesh <ben@benlesh.com>, Domenic Farolino <dom@chromium.org>
```
Comment (domfarolino) [inline LICENSE.txt:3]: Dominic. Not Domenic

**Comment by domfarolino** [comment]
See https://github.com/WICG/observable/pull/42#issuecomment-1689972876.

File: impl/observable-polyfill.d.ts [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+		on<K extends keyof AbortSignalEventMap>(
+			eventType: K,
+		): Observable<AbortSignalEventMap[K]>;
+	}
+	interface AbstractWorker {
+		on<K extends keyof AbstractWorkerEventMap>(
+			eventType: K,
+		): Observable<AbstractWorkerEventMap[K]>;
+	}
+	interface Animation {
```
Comment (domfarolino) [inline impl/observable-polyfill.d.ts:17]: Why are all of these being added?

File: interfaces.txt [hunk 5]
```diff
@@ -0,0 +1,73 @@
+AbortSignal
```
Comment (domfarolino) [inline interfaces.txt:1]: What is this file?

File: impl/observable-polyfill.d.ts (hunk 4 above)
Comment (benlesh) [reply impl/observable-polyfill.d.ts:17]: Honestly? Just to be nice. If anyone wants to use this with TypeScript, these types will help them. They don't *need* to be here.

File: interfaces.txt (hunk 5 above)
Comment (benlesh) [reply interfaces.txt:1]: Cruft. Sorry.

File: LICENSE.txt (hunk 3 above)
Comment (benlesh) [reply LICENSE.txt:3]: Sorry... 😬

File: impl/observable-polyfill.d.ts (hunk 4 above)
Comment (benlesh) [reply impl/observable-polyfill.d.ts:17]: My two cents is we leave them. They can't do anything but help folks use this and try it out.

Comment (keithamus) [reply impl/observable-polyfill.d.ts:17]: They map the event name strings to the event subclasses. This way `eventTarget.on('click')` will typed as an `Observable<MouseEvent>` while `eventTarget.on('keypress')` will be typed as an `Observable<KeyboardEvent>`.

File: impl/observable.js [hunk 6]
```diff
@@ -0,0 +1,650 @@
...
+	#abortController = new AbortController();
+	#signal;
+	#teardowns;
+	get signal() {
+		return this.#signal;
+	}
+	get isActive() {
+		return this.#active && !this.signal.aborted;
```
Comment (keithamus) [inline impl/observable.js:561]: This doesn't seem to align to the current spec. AFAICT it's possible to see `isActive === false` while also `this.signal.aborted === false`.

File: impl/observable.js [hunk 7]
```diff
@@ -0,0 +1,650 @@
...
+				? { next: fnOrObserver }
+				: fnOrObserver;
+		const observer = {
+			next: noop,
+			error: reportError,
+			complete: noop,
+			...partialObserver,
+		};
+		const subscriber = new Subscriber(options?.signal, observer);
+		this.start(subscriber);
```
Comment (keithamus) [inline impl/observable.js:74]: Does this need to try/catch?

**Comment by domfarolino** [comment]
This PR really hasn't seen any activity in months, and with Chromium getting pretty close to shipping Observables, I feel like we should probably just close this PR out. But feel free to re-open if you have plans to get to it.

**Comment by dariomannu** [comment]
Hey, shame this has been closed, as it was a good start.

OK that Chrome has it built-in, but it's still feature-flagged and other browsers don't have it at all.
Any strong opinions against reviving this and having an "official" polyfill? I'm happy to update this to the current spec, package it up, help co-maintain it, create some examples, tests, stackblitzes, whatever...

**Pull request closed without merging**